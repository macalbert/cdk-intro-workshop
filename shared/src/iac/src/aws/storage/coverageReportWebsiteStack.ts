import { CfnOutput, Duration, RemovalPolicy } from "aws-cdk-lib";
import { Certificate } from "aws-cdk-lib/aws-certificatemanager";
import {
    Distribution,
    type ErrorResponse,
    OriginAccessIdentity,
    ViewerProtocolPolicy,
} from "aws-cdk-lib/aws-cloudfront";
import { S3BucketOrigin } from "aws-cdk-lib/aws-cloudfront-origins";
import {
    ARecord,
    PublicHostedZone,
    RecordTarget,
} from "aws-cdk-lib/aws-route53";
import { CloudFrontTarget } from "aws-cdk-lib/aws-route53-targets";
import {
    BlockPublicAccess,
    Bucket,
    BucketAccessControl,
    BucketEncryption,
    HttpMethods,
} from "aws-cdk-lib/aws-s3";
import type { Construct } from "constructs";
import { M47Stack, type M47StackProps } from "../../config/shared/m47Stack";

/**
 * Properties for configuring the CoverageReportWebsiteStack.
 *
 * This interface extends M47StackProps and includes properties required to deploy a static website
 * for code coverage reports. The properties include:
 * - bucketName: The name of the S3 bucket to host the website.
 * - subdomainName: The subdomain for the website.
 * - certificateArn: The ARN of the ACM certificate used for the CloudFront distribution.
 * - hostedZoneId: The ID of the Route 53 hosted zone.
 * - domain: The domain name for the website.
 *
 * @example
 * const props: CoverageReportWebsiteStackProps = {
 *   bucketName: "coverage-report-bucket",
 *   subdomainName: "coverage",
 *   certificateArn: "arn:aws:acm:region:account:certificate/abc123",
 *   hostedZoneId: "Z1234567890",
 *   domain: "example.com",
 *   // plus additional M47StackProps properties...
 * };
 */
export interface CoverageReportWebsiteStackProps extends M47StackProps {
    bucketName: string;
    subdomainName: string;
    certificateArn: string;
    hostedZoneId: string;
    domain: string;
}

/**
 * CoverageReportWebsiteStack provisions a static website for code coverage reports using S3 and CloudFront.
 *
 * The stack creates an S3 bucket to securely host the static website and configures a CloudFront distribution
 * with custom error responses. It also sets up a Route 53 A record for the custom domain, allowing the website
 * to be accessed via a friendly URL.
 *
 * The CloudFront distribution is configured with error responses for HTTP 403 and 404 that both redirect to
 * index.html. Project-specific tags are applied to resources, and CloudFormation outputs provide the CloudFront
 * distribution domain name and the DNS record.
 *
 * @example
 * new CoverageReportWebsiteStack(app, {
 *   bucketName: "coverage-report-bucket",
 *   subdomainName: "coverage",
 *   certificateArn: "arn:aws:acm:region:account:certificate/abc123",
 *   hostedZoneId: "Z1234567890",
 *   domain: "example.com",
 *   // plus additional M47StackProps properties...
 * });
 */
export class CoverageReportWebsiteStack extends M47Stack {
    constructor(scope: Construct, props: CoverageReportWebsiteStackProps) {
        super(scope, props);

        // Construct the full domain name (e.g., coverage.example.com)
        const fullDomainName = [props.subdomainName, props.domain]
            .join(".")
            .toLowerCase();

        // Create a private S3 bucket for hosting the static website
        const bucketWebsite = new Bucket(this, "WebsiteBucket", {
            accessControl: BucketAccessControl.PRIVATE,
            publicReadAccess: false,
            versioned: false,
            removalPolicy: RemovalPolicy.DESTROY,
            bucketName: props.bucketName,
            autoDeleteObjects: true,
            blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
            encryption: BucketEncryption.S3_MANAGED,
            cors: [
                {
                    allowedMethods: [HttpMethods.GET, HttpMethods.HEAD],
                    allowedOrigins: ["*"],
                    allowedHeaders: ["*"],
                },
            ],
            enforceSSL: true,
        });

        // Apply project-specific tags to the bucket
        this.addProjectTags(bucketWebsite, props);

        // Create an Origin Access Identity for CloudFront to securely access the S3 bucket
        const originAccessIdentity = new OriginAccessIdentity(
            this,
            "originAccessIdentity",
            {
                comment: `Setup access from CloudFront to the bucket ${fullDomainName} (read)`,
            },
        );

        // Grant read access to the Origin Access Identity
        bucketWebsite.grantRead(originAccessIdentity);

        // Import the ACM certificate for the custom domain
        const certificate = Certificate.fromCertificateArn(
            this,
            "Certificate",
            props.certificateArn,
        );

        // Define custom error responses for the CloudFront distribution
        const errorResponses: ErrorResponse[] = [
            {
                httpStatus: 403,
                responseHttpStatus: 200,
                responsePagePath: "/index.html",
                ttl: Duration.seconds(10),
            },
            {
                httpStatus: 404,
                responseHttpStatus: 200,
                responsePagePath: "/index.html",
                ttl: Duration.seconds(10),
            },
        ];

        // Create the CloudFront distribution for the website
        const distribution = new Distribution(this, "distribution", {
            domainNames: [fullDomainName],
            defaultBehavior: {
                origin: S3BucketOrigin.withOriginAccessIdentity(bucketWebsite, {
                    originAccessIdentity: originAccessIdentity,
                }),
                viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
            },
            defaultRootObject: "index.html",
            certificate: certificate,
            errorResponses: errorResponses,
        });

        // Apply project-specific tags to the CloudFront distribution
        this.addProjectTags(bucketWebsite, props);

        // Import the Route 53 hosted zone using provided attributes
        const zoneFromAttributes =
            PublicHostedZone.fromPublicHostedZoneAttributes(
                this,
                "publicHostedZone",
                {
                    zoneName: props.domain,
                    hostedZoneId: props.hostedZoneId,
                },
            );

        // Create a Route 53 A record mapping the full domain name to the CloudFront distribution
        const aliasRecord = new ARecord(this, "AliasRecord", {
            recordName: fullDomainName,
            target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
            zone: zoneFromAttributes,
        });

        // Apply project-specific tags to the alias record
        this.addProjectTags(aliasRecord, props);

        // Output the CloudFront distribution domain name for external reference
        new CfnOutput(this, "CoverageReportCdn", {
            value: distribution.distributionDomainName,
            description:
                "The domain name of the CloudFront distribution for Coverage Report",
            exportName: `${this.toCloudFormation()}CoverageReportCdn`.replace(
                /-/g,
                "",
            ),
        });

        // Output the DNS record name for external reference
        new CfnOutput(this, "CoverageReportDns", {
            value:
                aliasRecord.domainName ||
                `${fullDomainName}.${zoneFromAttributes.zoneName}`,
            description: "The domain name for the alias record",
            exportName: `${this.toCloudFormation()}CoverageReportDns`.replace(
                /-/g,
                "",
            ),
        });
    }
}
