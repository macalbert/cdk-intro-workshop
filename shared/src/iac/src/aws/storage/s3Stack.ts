import { RemovalPolicy } from "aws-cdk-lib";
import { Bucket, HttpMethods } from "aws-cdk-lib/aws-s3";
import type { Construct } from "constructs";
import { M47Stack, type M47StackProps } from "../../config/shared/m47Stack";

/**
 * Properties for configuring the S3Stack.
 *
 * This interface extends M47StackProps and includes a bucketName property which defines the name of the S3 bucket.
 *
 * @example
 * const props: S3StackProps = {
 *   bucketName: "my-bucket",
 *   // plus additional M47StackProps properties...
 * };
 */
export interface S3StackProps extends M47StackProps {
    bucketName: string;
}

/**
 * S3Stack provisions an S3 bucket with specific configuration.
 *
 * The bucket is created using the provided bucketName with a RETAIN removal policy and SSL enforcement enabled.
 * Additionally, a CORS rule is added to allow GET, PUT, and HEAD requests from any origin.
 *
 * @example
 * new S3Stack(app, {
 *   bucketName: "my-bucket",
 *   // plus additional M47StackProps properties...
 * });
 */
export class S3Stack extends M47Stack {
    public readonly bucket: Bucket;

    /**
     * Constructs a new instance of the S3Stack.
     *
     * This constructor creates an S3 bucket using the specified bucketName and configuration.
     * It sets a RETAIN removal policy, enforces SSL, and adds a CORS rule to allow specific HTTP methods.
     *
     * @param scope - The scope in which this stack is defined.
     * @param props - The properties for configuring the S3 bucket.
     */
    constructor(scope: Construct, props: S3StackProps) {
        super(scope, props);

        this.bucket = new Bucket(this, "S3Bucket", {
            bucketName: props.bucketName,
            removalPolicy: RemovalPolicy.RETAIN,
            enforceSSL: true,
        });

        this.bucket.addCorsRule({
            allowedOrigins: ["*"],
            allowedMethods: [
                HttpMethods.GET,
                HttpMethods.PUT,
                HttpMethods.HEAD,
            ],
            allowedHeaders: ["*"],
        });
    }
}
