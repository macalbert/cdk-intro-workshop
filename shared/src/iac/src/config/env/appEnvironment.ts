export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
