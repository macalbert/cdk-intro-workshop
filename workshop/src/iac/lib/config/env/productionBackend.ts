import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
