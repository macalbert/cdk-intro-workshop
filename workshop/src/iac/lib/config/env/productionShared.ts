import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
