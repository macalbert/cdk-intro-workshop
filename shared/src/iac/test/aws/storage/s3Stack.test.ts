import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { S3Stack, type S3StackProps } from "../../../src/aws/storage/s3Stack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("S3Stack", () => {
	const env = {
		account: "account",
		region: "eu-west-1",
	};

	test("Should_CreateBucket_When_StackIsCalled", () => {
		// Arrange
		const stack = new Stack(new App(), "s3Stack", {
			env: env,
		});
		const props = createS3StackProps();

		// Act
		const actual = new S3Stack(stack, props);

		// Assert
		const template = Template.fromStack(actual);
		expect(template.toJSON()).toMatchSnapshot("s3Stack");
	});

	function createS3StackProps(): S3StackProps {
		return {
			env: env,
			bucketName: "bucket-test",
			githubRepo: "Test",
			envName: AppEnvironment.Production,
			name: "test-CoverageReport",
			stackName: "test-CoverageReportStack",
		};
	}
});
