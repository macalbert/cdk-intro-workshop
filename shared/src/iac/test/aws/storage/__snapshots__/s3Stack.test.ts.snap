// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3Stack Should_CreateBucket_When_StackIsCalled: s3Stack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "S3Bucket07682993": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "bucket-test",
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
                "PUT",
                "HEAD",
              ],
              "AllowedOrigins": [
                "*",
              ],
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "S3BucketPolicyF560589A": {
      "Properties": {
        "Bucket": {
          "Ref": "S3Bucket07682993",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3Bucket07682993",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3Bucket07682993",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
